src/gui/qgsmediawidget.cpp

```diff
@@ -0,0 +1,197 @@
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsmediawidget.h"
+#include "qgsapplication.h"
+
+#include <QLabel>
+#include <QHBoxLayout>
+#include <QVBoxLayout>
+#include <QPushButton>
+#include <QSlider>
+#include <QVideoWidget>
+
+
+QgsMediaWidget::QgsMediaWidget( QWidget *parent )
+  : QWidget( parent )
+{
+  mLayout = new QVBoxLayout();
+  mLayout->setContentsMargins( 0, 0, 0, 0 );
+
+  mVideoWidget = new QVideoWidget( this );
+  mVideoWidget->setSizePolicy( QSizePolicy::Expanding, QSizePolicy::Expanding );
+  mVideoWidget->setMinimumHeight( 100 );
+  mLayout->addWidget( mVideoWidget );
+
+  QHBoxLayout *controlsLayout = new QHBoxLayout();
+  controlsLayout->setContentsMargins( 0, 0, 0, 0 );
+
+  mPlayButton = new QPushButton( this );
+  mPlayButton->setSizePolicy( QSizePolicy::Minimum, QSizePolicy::Preferred );
+  mPlayButton->setIcon( QgsApplication::getThemeIcon( QStringLiteral( "/mActionPlay.svg" ) ) );
+  mPlayButton->setCheckable( true );
+  controlsLayout->addWidget( mPlayButton );
+
+  mPositionSlider = new QSlider( this );
+  mPositionSlider->setSizePolicy( QSizePolicy::Expanding, QSizePolicy::Preferred );
+  mPositionSlider->setOrientation( Qt::Horizontal );
+  controlsLayout->addWidget( mPositionSlider );
+
+  mDurationLabel = new QLabel( this );
+  mDurationLabel->setSizePolicy( QSizePolicy::Minimum, QSizePolicy::Preferred );
+  mDurationLabel->setAlignment( Qt::AlignHCenter );
+  mDurationLabel->setText( QStringLiteral( "-" ) );
+  QFontMetrics fm( mDurationLabel->font() );
+  mDurationLabel->setMinimumWidth( fm.boundingRect( QStringLiteral( "00:00:00" ) ).width() );
+  controlsLayout->addWidget( mDurationLabel );
+
+  QWidget *controls = new QWidget();
+  controls->setLayout( controlsLayout );
+
+  mLayout->addWidget( controls );
+  setLayout( mLayout );
+
+  adjustControls();
+  toggleControls( false );
+
+  mMediaPlayer.setVideoOutput( mVideoWidget );
+
+  connect( &mMediaPlayer, &QMediaPlayer::mediaStatusChanged, this, &QgsMediaWidget::mediaStatusChanged );
+  connect( &mMediaPlayer, &QMediaPlayer::positionChanged, this, [ = ]()
+  {
+    mPositionSlider->setValue( mMediaPlayer.position() );
+  } );
+
+  connect( mPlayButton, &QAbstractButton::clicked, this, [ = ]()
+  {
+    if ( mMediaPlayer.state() == QMediaPlayer::PlayingState )
+    {
+      mMediaPlayer.pause();
+    }
+    else
+    {
+      mMediaPlayer.play();
+    }
+  } );
+  connect( mPositionSlider, &QAbstractSlider::sliderReleased, this, [ = ]()
+  {
+    mMediaPlayer.setPosition( mPositionSlider->value() );
+  } );
+}
+
+void QgsMediaWidget::setMediaPath( const QString &path )
+{
+  if ( mMediaPath == path )
+    return;
+
+  mMediaPath = path;
+  mMediaPlayer.setMedia( QUrl::fromLocalFile( path ) );
+
```

Comment:
It would be better if the proper `setMode()` was called depending on `mMediaPlayer.isVideoAvailable()`

Reply:
@uclaros , I disagree here, you might actually want to consume a video file as audio-only.

Reply:
@nirvn Is the video widget visible when playing audio files in video mode?
If yes, you could at least do something like 
```cpp
if ( mMode == Video )
  mVideoWidget->setVisible( mMediaPlayer.isVideoAvailable() );
```